gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageBidiReadChannel.java

```diff
@@ -0,0 +1,151 @@
+      Storage storage,
+      GoogleCloudStorageItemInfo itemInfo,
+      GoogleCloudStorageReadOptions readOptions,
+      ExecutorService boundedThreadPool)
+      throws IOException {
+    this.storage = storage;
+    this.resourceId =
+        new StorageResourceId(
+            itemInfo.getBucketName(), itemInfo.getObjectName(), itemInfo.getContentGeneration());
+    this.blobId =
+        BlobId.of(
+            resourceId.getBucketName(), resourceId.getObjectName(), resourceId.getGenerationId());
+    this.readOptions = readOptions;
+    this.blobReadSession =
+        initializeBlobReadSession(storage, blobId, readOptions.getBidiClientTimeout());
+    this.boundedThreadPool = boundedThreadPool;
+  }
+
+  private BlobReadSession initializeBlobReadSession(
+      Storage storage, BlobId blobId, int clientTimeout) throws IOException {
+    try {
+      return storage.blobReadSession(blobId).get(clientTimeout, TimeUnit.SECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public int read(ByteBuffer dst) throws IOException {
+    // TODO(dhritichopra) Add read flow
+    return 0;
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    throw new UnsupportedOperationException("Cannot mutate read-only channel");
+  }
+
+  @Override
+  public long position() throws IOException {
+    // TODO(dhritichopra) Add read flow
+    return 0;
+  }
+
+  @Override
+  public SeekableByteChannel position(long newPosition) throws IOException {
+    // TODO(dhritichopra) Add read flow
+    return null;
+  }
+
+  @Override
+  public long size() throws IOException {
+    // TODO(dhritichopra) Add read flow
+    return 0;
+  }
+
+  @Override
+  public SeekableByteChannel truncate(long size) throws IOException {
+    // TODO(dhritichopra) Add read flow
+    return null;
+  }
+
+  @Override
+  public boolean isOpen() {
+    // TODO(dhritichopra) Add read flow
+    return false;
+  }
+
+  @Override
+  public void close() throws IOException {
+    blobReadSession.close();
+  }
+
+  @Override
+  public void readVectored(List<VectoredIORange> ranges, IntFunction<ByteBuffer> allocate)
+      throws IOException {
+    ranges.forEach(
+        range -> {
+          ApiFuture<DisposableByteString> futureBytes =
+              blobReadSession.readAs(
+                  ReadProjectionConfigs.asFutureByteString()
+                      .withRangeSpec(RangeSpec.of(range.getOffset(), range.getLength())));
+          ApiFutures.transform(
+              futureBytes,
+              disposableByteString -> {
+                try {
+                  return processBytesAndCompleteRange(disposableByteString, range, allocate);
+                } catch (Throwable t) {
+                  range.getData().completeExceptionally(t);
+                  return null;
+                }
+              },
+              boundedThreadPool);
+        });
```

Comment:
You'll want to change this from `transform` instead to `addCallback` so that you can observe if the future fails. `transform` will only apply to a successful resolution, and is generally useful when you want to apply a modification to a future you intend to return for awaiting elsewhere.

```suggestion
          ApiFutures.addCallback(
              futureBytes,
              new ApiFutureCallback<>() {
                @Override
                public void onFailure(Throwable t) {
                  range.getData().completeExceptionally(t);
                }

                @Override
                public void onSuccess(DisposableByteString disposableByteString) {
                  try {
                    processBytesAndCompleteRange(disposableByteString, range, allocate);
                  } catch (Throwable t) {
                    range.getData().completeExceptionally(t);
                  }
                }
              },
              boundedThreadPool);
```

Reply:
Done